breakpad-symbols/src/sym_file/mod.rs

```diff
@@ -69,267 +69,182 @@ pub mod walker;
-                    // (do nothing, fallthrough to normal path).
-                } else if fully_consumed {
-                    // Success! The last iteration cleared the buffer and we still got
-                    // no more bytes, so that's a proper EOF with a complete parse!
-                    return Ok(parser.finish());
-                } else if !tried_to_grow {
-                    // We still have some stuff in the buffer, assume this is because
-                    // the buffer is full, and try to make it BIGGER and ask for more again.
-                    let new_cap = buf.capacity().saturating_mul(2);
-                    if new_cap > MAX_BUFFER_CAPACITY {
-                        // TIME TO PANIC!!! This line is catastrophically big, just start
-                        // discarding bytes until we hit a newline.
-                        trace!("RECOVERY: discarding enormous line {}", parser.lines);
-                        in_panic_recovery = true;
-                        continue;
-                    }
-                    trace!("parser out of space? trying more ({}KB)", new_cap / 1024);
-                    buf.grow(new_cap);
-                    tried_to_grow = true;
-                    continue;
-                } else if total_consumed == 0 {
-                    // We grew the buffer and still got no more bytes, so it's a proper EOF.
-                    // But actually, we never consumed any bytes, so this is an empty file?
-                    // Give a better error message for that.
-                    return Err(SymbolError::ParseError(
-                        "empty SymbolFile (probably something wrong with your debuginfo tooling?)",
-                        0,
-                    ));
-                } else {
-                    // Ok give up, this input is just impossible.
-                    return Err(SymbolError::ParseError(
-                        "unexpected EOF during parsing of SymbolFile (or a line was too long?)",
-                        parser.lines,
-                    ));
-                }
-            } else {
-                tried_to_grow = false;
-            }
-
-            if in_panic_recovery {
-                // Don't run the normal parser while we're still recovering!
-                continue;
-            }
-            just_finished_recovering = false;
-
-            // Ask the parser to parse more of the input
-            let input = buf.data();
-            let consumed = parser.parse_more(input)?;
-            total_consumed += consumed as u64;
-
-            // Give the other consumer of this Reader a chance to use this data.
-            callback(&input[..consumed]);
+pub struct StreamingSymbolParser {
+    buf: circular::Buffer,
+    parser: SymbolParser,
+
+    fully_consumed: bool,
+    tried_to_grow: bool,
+    in_panic_recovery: bool,
+    just_finished_recovering: bool,
+    total_consumed: u64,
+}
```

Comment:
All of this stuff is just the existing parse_async code but stateful and with an `add_input` and `finish` method so that an external source can handle getting bytes and drive the streaming parser, without it actually being async. This means all the sync and async paths are now unified and all the callback crap is now gone.